### PR TITLE
comp: Free model_handler in error paths during create

### PR DESCRIPTION
**Pull request by afq984**
For #4342 

Add matching comp_data_blob_handler_free calls for
comp_data_blob_handler_new in error paths, which are
missing in many components' "new" functions.

Signed-off-by: Li-Yu Yu <aaronyu@google.com>

### CODE REVIEW

File: src/audio/igo_nr/igo_nr.c
Comment (cujomalainey) [inline src/audio/igo_nr/igo_nr.c:337]: typically the method here is to walk things out in reverse so you don't have `if` and instead have multiple labels, then adjust goto's as neccessary
e.g.
```
comp_err:
  comp_data_blob_handler_free(cd->model_handler);
err:
  rfree(cd);
  rfree(dev);
  return NULL;
```

File: src/samples/audio/smart_amp_test.c
Comment (cujomalainey) [inline src/samples/audio/smart_amp_test.c:83]: rfree(NULL) is a no-op so you don't have to check for nullity.

File: src/samples/audio/detect_test.c
Comment (cujomalainey) [inline src/samples/audio/detect_test.c:337]: lets remove this if with same method as above.

**Review by cujomalainey — CHANGES_REQUESTED** [review]
I'm noticing a lot of the components don't use goto's to exit and its leading to multiple long exits. While we are here we should fix that. Thanks this is a great start.

File: src/samples/audio/smart_amp_test.c
Comment (afq984) [reply src/samples/audio/smart_amp_test.c:83]: `if (sad)` was intended to guard `sad->model_handler`. Currently all jumps to `fail` has `sad != NULL` so yes this check is unnecessary.

**Review by lgirdwood — COMMENTED** [review]
Good find @afq984

File: src/samples/audio/smart_amp_test.c
Comment (cujomalainey) [reply src/samples/audio/smart_amp_test.c:83]: Yep, that would be bad, but that is where the second label comes in like above to flatten out the code.

File: src/audio/drc/drc.c
Comment (cujomalainey) [inline src/audio/drc/drc.c:164]: same here since it is only the first alloc

File: src/audio/drc/drc.c
Comment (cujomalainey) [inline src/audio/drc/drc.c:159]: sorry for not being clear, but first alloc can stay as a return since it does not have to free anything else.

**Review by cujomalainey — COMMENTED** [review]
Looks great just a little bit of over optimization, once that is fixed it looks good for submission

File: src/audio/drc/drc.c
Comment (afq984) [reply src/audio/drc/drc.c]: done, exit paths:
* `return NULL` = dev not allocated
* `goto fail` = dev allocated
* `goto cd_fail` = cd allocated

**Review by cujomalainey — APPROVED** [review]

**Comment by cujomalainey** [comment]
Will merge if we dont get a second review by EOW

File: src/audio/drc/drc.c
Comment (ranj063) [inline src/audio/drc/drc.c:179]: why do you need to go to cd_fail in this case? cd->model_handler was never successfully alloc'ed anyway right?

Comment (afq984) [reply src/audio/drc/drc.c:179]: `cd` still needs to be freed; `comp_data_blob_handler_free(NULL)` is no-op.
(My mental model for `cd_fail` was: jump here if cd is allocated)

Or we can move `rfree(cd)` below `fail:` instead.

Comment (cujomalainey) [reply src/audio/drc/drc.c:179]: @ranj063 `free(NULL)` handlers are no-ops and since `cd was allocated with `rzalloc` it is `NULL`. So this is just a matter of flavour. Label could be above `cd` but would serve same purpose.

**Review by lgirdwood — APPROVED** [review]

**Pull request merged**
